test/lib/lineage-constraints.spec.js

```diff
@@ -0,0 +1,139 @@
+PouchDB.plugin(require('pouchdb-adapter-memory'));
+PouchDB.plugin(require('pouchdb-mapreduce'));
+
+const { mockHierarchy } = require('../mock-hierarchies');
+
+const lineageConstraints = rewire('../../src/lib/lineage-constraints');
+const log = require('../../src/lib/log');
+log.level = log.LEVEL_INFO;
+
+describe('lineage constriants', () => {
+  describe('getConfigurableHierarchyErrors', () => {
+    const scenario = async (contact_types, contactType, parentType) => {
+      const mockDb = { get: () => ({ settings: { contact_types } }) };
+      const { getConfigurableHierarchyErrors } = await lineageConstraints(mockDb, { type: parentType });
+      return getConfigurableHierarchyErrors({ type: contactType });
+    };
+
+    it('no defined rules yields no error', async () => expect(await scenario(undefined, 'person', 'health_center')).to.be.undefined);
+
+    it('no defined rules yields not defined', async () => expect(await scenario([], 'person', 'health_center')).to.include('not define'));
+
+    it('no valid parent yields not defined', async () => expect(await scenario([undefined], 'person', 'health_center')).to.include('not define'));
+
+    it('valid parent yields no error', async () => {
+      const actual = await scenario([{
+        id: 'person',
+        parents: ['health_center'],
+      }], 'person', 'health_center');
+
+      expect(actual).to.be.undefined;
+    });
+
+    it('no contact type yields undefined error', async () => expect(await scenario([])).to.include('undefined'));
+
+    it('no parent type yields undefined error', async () => expect(await scenario([], 'person')).to.include('undefined'));
+
+    it('no valid parents yields not defined', async () => expect(await scenario([{
+      id: 'person',
+      parents: ['district_hospital'],
+    }], 'person', 'health_center')).to.include('does not allow parent'));
+
+    it('no settings doc requires valid parent type', async () => {
+      const mockDb = { get: () => { throw { name: 'not_found' }; } };
+      const { getConfigurableHierarchyErrors } = await lineageConstraints(mockDb, { type: 'dne' });
+      const actual = getConfigurableHierarchyErrors({ type: 'person' });
+      expect(actual).to.include('is not a contact');
+    });
+
+    it('no settings doc requires valid contact type', async () => {
+      const mockDb = { get: () => { throw { name: 'not_found' }; } };
+      const { getConfigurableHierarchyErrors } = await lineageConstraints(mockDb, { type: 'clinic' });
+      const actual = getConfigurableHierarchyErrors({ type: 'dne' });
+      expect(actual).to.include('is not a contact');
+    });
+
+    it('no settings doc yields not defined', async () => {
+      const mockDb = { get: () => { throw { name: 'not_found' }; } };
+      const { getConfigurableHierarchyErrors } = await lineageConstraints(mockDb, { type: 'clinic' });
+      const actual = getConfigurableHierarchyErrors({ type: 'person' });
+      expect(actual).to.be.undefined;
+    });
```

Comment:
Consider adding tests that prove you cannot create contacts that don't follow the default contact schema.

Reply:
Done